[webkitscmpy] Add method to retrieve last N commits for a specific file

karlrackler opened the pull request:
#### 875467dc29efa0c369307f968fbbd1c8c23843aa
<pre>
[webkitscmpy] Add method to retrieve last N commits for a specific file
<a href="https://bugs.webkit.org/show_bug.cgi?id=290338">https://bugs.webkit.org/show_bug.cgi?id=290338</a>
<a href="https://rdar.apple.com/147777529">rdar://147777529</a>

Reviewed by Jonathan Bedard.

Add a new method last_commits_on to the Git class in webkitscmpy to return the last N commits that modified

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
(Git.last_commits_on):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git.last_commits_on):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/git_unittest.py:

Canonical link: <a href="https://commits.webkit.org/294287@main">https://commits.webkit.org/294287@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43bfd333af7578258e1e4e12724b86591ef92b40

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101356 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21020 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11325 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106509 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51985 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103396 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21328 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29514 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/77199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34236 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104363 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/16458 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91551 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/57540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/100829 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16278 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9561 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51333 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86159 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9637 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108861 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28485 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/100913 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28847 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87753 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/85739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21809 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/30459 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/8170 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22602 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28416 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28227 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31547 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29786 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
```diff
@@ -991,6 +991,19 @@ def commits(self, begin=None, end=None, include_log=True, include_identifier=Tru
+        """
+        try:
+            output = subprocess.check_output(
+                ['git', '-C', self.root_path, 'log', f'--max-count={count}', '--follow', '--format=%H', '--', path],
```
Change `git` to `self.executable()`.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/git_unittest.py:
```diff
@@ -614,6 +614,57 @@ def test_is_suitable_branch_for_pull_request(self):
             self.assertEqual(repo.is_suitable_branch_for_pull_request('branch-a', source_remote=source_remote), False)
             self.assertEqual(repo.is_suitable_branch_for_pull_request(None, source_remote=source_remote), False)
 
+    def test_last_commits_on(self):
```
This test should be most similar to `test_files_changed`. You'll just need the `with mocks.local.Git(self.path), OutputCapture():` as your mock, `local.last_commits_on` won't "just work", though, you're going to need to mock the `git` command you're calling in `Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py`

karlrackler replied:
Okay I updated following your directions.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
```diff
@@ -26,7 +26,7 @@
 import time
 from collections import OrderedDict
-from datetime import datetime, timezone
+from datetime import datetime
```
I think this is your only issue! If you re-add the `, timezone` import, I think all the tests will pass.

karlrackler replied:
I just did.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
```diff
@@ -991,6 +991,23 @@ def commits(self, begin=None, end=None, include_log=True, include_identifier=Tru
+                for commit_hash in output.stdout.strip().splitlines()
+                if commit_hash.strip()
+            ]
+        except Exception:
```
We should catch a more narrow exception here. Specifically, I believe `subprocess.CalledProcessError` is the right exception to be catching. (alternatively, you could just check the exit code of `output` and return `[]` if it's non-zero)

karlrackler committed:
[webkitscmpy] Add method to retrieve last N commits for a specific file
https://bugs.webkit.org/show_bug.cgi?id=290338
rdar://147777529

Reviewed by Jonathan Bedard.

Add a new method last_commits_on to the Git class in webkitscmpy to return the last N commits that modified

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
(Git.last_commits_on):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git.last_commits_on):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/git_unittest.py:

Canonical link: https://commits.webkit.org/294287@main
